drivers: i2s: nrfx: Allow MCK bypass for proper LRCK and MCK Ratios

standalone-tech opened the pull request:
nRF53 series SoCs have a dedicated configurable audio PLL and the ability to enable MCK bypass via a register value CONFIG.CLKCONFIG. This can enable higher MCK/LRCK ratios that some I2S peripherals require the host to generate. Allow an application developer to choose if they want to initially look for a bypass ratio and, if found, enable bypass in the NRFX driver. If not, the standard MCK calculation is conducted as normal.

standalone-tech committed:
drivers: i2s: nrfx: Allow MCK bypass for proper LRCK and MCK Ratios

nRF53 series SoCs have a dedicated configurable audio PLL and the ability
to enable MCK bypass via a register value CONFIG.CLKCONFIG. This can
enable higher MCK/LRCK ratios that some I2S peripherals require the host
to generate. Allow an application developer to choose if they want to
initially look for a bypass ratio and, if found, enable bypass in the
NRFX driver. If not, the standard MCK calculation is conducted as normal.

Signed-off-by: Sean O'Connor <sean@standalone.tech>

tomi-font commented:
All reviewers (and especially @anangl): please review

anangl commented on drivers/i2s/i2s_nrfx.c:
```diff
@@ -95,6 +95,26 @@ static void find_suitable_clock(const struct i2s_nrfx_drv_cfg *drv_cfg,
+			best_mck = src_freq;
+			best_diff = 0;
+
+			best_mck_cfg = NRF_I2S_MCK_32MDIV8;
```
What's the point in setting this? The value is written to the `CONFIG.MCKFREQ` register, which is ignored when the bypass is enabled.

anangl commented on drivers/i2s/i2s_nrfx.c:
```diff
@@ -95,6 +95,26 @@ static void find_suitable_clock(const struct i2s_nrfx_drv_cfg *drv_cfg,
+			break;
+		}
+	}
+	if (best_diff) {
+		config->enable_bypass = false;
+	}
```
This is redundant. `enable_bypass` is set to `false` by default: https://github.com/zephyrproject-rtos/zephyr/blob/0a43c41b486c2e0274577153e3eaf520d53bbc3d/drivers/i2s/i2s_nrfx.c#L472

standalone-tech replied:
Should the whole if statement be removed?

standalone-tech replied:
Removed

anangl commented on samples/drivers/i2s/echo/boards/nrf5340dk_nrf5340_cpuapp.overlay:
```diff
@@ -22,7 +22,8 @@
 	i2s0_default_alt: i2s0_default_alt {
 		group1 {
-			psels = <NRF_PSEL(I2S_SCK_M, 1, 15)>,
+			psels = <NRF_PSEL(I2S_MCK, 0, 12)>,
```
Why?

standalone-tech committed:
samples: i2s: echo: Enable nRF53 MCK output in I2S sample

Enable the MCK output to a pin to allow for DAC and ADC clocking.
Provides a MCK source to peripherals that can utilize an external
clocking source for internal DSP processes, especially at higher
MCK/LRCK ratios.

Signed-off-by: Sean O'Connor <sean@standalone.tech>

anangl reviewed (dismissed):
The title of the second commit should start with something like `samples: i2s: echo:` instead of just `Sample:`.

standalone-tech commented:
> > @anangl Is there something I should be doing about the twister python library failure? I don't know what I might've done to cause that
> > ```
> > ModuleNotFoundError: No module named 'jsonschema'
> > ```
> 
> Please rebase your branch. This should solve the problem.

Rebased and pushed
